### PR TITLE
fix: add annotations and enable nullable check

### PR DESCRIPTION
**Pull request by ghost**
This is a fix to enable nullable check for the
"Microsoft.Extensions.Configuration.Abstractions"
by enabling using netcoreapp3.1 and the latest version
of .net core, including:

1. Enable "Nullable".
2. Change some missing / wrong comments in the XML document.

Ref: https://github.com/dotnet/runtime/issues/57200

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57267) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57267) before we can accept your contribution.<br/><br/>:x: MaledongGit [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=57267)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=57267) it.</sub>

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs [hunk 1]
```diff
@@ -35,7 +36,7 @@ public static string Combine(params string[] pathSegments)
         /// </summary>
         /// <param name="pathSegments">The path segments to combine.</param>
         /// <returns>The combined path.</returns>
-        public static string Combine(IEnumerable<string> pathSegments)
+        public static string Combine([NotNull]IEnumerable<string> pathSegments)
```
Comment (maxkoshevoi) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:39]: Why did you need to specify this? Isn't `pathSegments` not null by default?

Comment (MaledongGit) [reply src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationPath.cs:39]: Thank you, I fixed it but not submitted in the Git Stage....

File: src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs [hunk 2]
```diff
@@ -28,10 +29,10 @@ public static IConfigurationBuilder Add<TSource>(this IConfigurationBuilder buil
         /// <summary>
         /// Shorthand for GetSection("ConnectionStrings")[name].
         /// </summary>
-        /// <param name="configuration">The configuration.</param>
+        /// <param name="configuration">The <see cref="IConfiguration"/> to enumerate.</param>
         /// <param name="name">The connection string key.</param>
         /// <returns>The connection string.</returns>
-        public static string GetConnectionString(this IConfiguration configuration, string name)
+        public static string? GetConnectionString(this IConfiguration configuration, string name)
```
Comment (maxkoshevoi) [inline src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/ConfigurationExtensions.cs:35]: I think you also need to do the same changes in `Ref` project

**Review by maxkoshevoi — CHANGES_REQUESTED** [review]
- `Exists` method should look like `public static bool Exists([NotNullWhen(true)] this IConfigurationSection? section)`

- `ConfigurationPath.GetParentPath` -> `public static string? GetParentPath(string? path)`
- `ConfigurationPath.GetSectionKey` -> `[return: NotNullIfNotNull("path")] public static string? GetSectionKey(string? path)`
- `ConfigurationRootExtensions.GetValueAndProvider` -> `private static (string? Value, IConfigurationProvider? Provider) GetValueAndProvider`
- `IConfiguration.this[]` -> `string? this[string key] { get; set; }`
- `IConfiguration.GetSection` -> `IConfigurationSection? GetSection(string key)`
- `IConfigurationProvider.TryGet` -> `bool TryGet(string key, [NotNullWhen(true)] out string? value)`

And there may be some other places that I've missed. Could you please annotate all implementations as well, to ensure that nullable attributes are correct?

**Comment by maxkoshevoi** [comment]
Could you give this PR to me please, since I've asked in the issue to implement it and already began the work?

**Comment by MaledongGit** [comment]
@maxkoshevoi: OK, I'll close it and you can re-open a new PR.
